spring-cloud-loadbalancer/src/main/java/org/springframework/cloud/loadbalancer/core/HintBasedServiceInstanceListSupplier.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2012-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.loadbalancer.core;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import reactor.core.publisher.Flux;
+
+import org.springframework.cloud.client.ServiceInstance;
+import org.springframework.cloud.client.loadbalancer.HintRequestContext;
+import org.springframework.cloud.client.loadbalancer.LoadBalancerProperties;
+import org.springframework.cloud.client.loadbalancer.Request;
+import org.springframework.cloud.client.loadbalancer.RequestDataContext;
+import org.springframework.http.HttpHeaders;
+import org.springframework.util.StringUtils;
+
+/**
+ * A {@link ServiceInstanceListSupplier} implementation that uses hints to filter service
+ * instances provided by the delegate.
+ *
+ * @author Olga Maciaszek-Sharma
+ * @since 3.0.2
+ */
+public class HintBasedServiceInstanceListSupplier extends DelegatingServiceInstanceListSupplier {
+
+	private final LoadBalancerProperties properties;
+
+	public HintBasedServiceInstanceListSupplier(ServiceInstanceListSupplier delegate,
+			LoadBalancerProperties properties) {
+		super(delegate);
+		this.properties = properties;
+	}
+
+	@Override
+	public Flux<List<ServiceInstance>> get() {
+		return delegate.get();
+	}
+
+	@Override
+	public Flux<List<ServiceInstance>> get(Request request) {
+		if (request.getContext() == null) {
+			return get();
+		}
+		if (request.getContext() instanceof RequestDataContext) {
+			RequestDataContext context = (RequestDataContext) request.getContext();
+			if (context.getClientRequest() != null) {
+				HttpHeaders headers = context.getClientRequest().getHeaders();
+				if (headers != null) {
+					String hintFromHeader = headers.getFirst(properties.getHintHeaderName());
+					if (StringUtils.hasText(hintFromHeader)) {
+						return get().map(instances -> filteredByHint(instances, hintFromHeader));
+					}
+				}
+			}
+		}
+		if (request.getContext() instanceof HintRequestContext) {
```

Comment:
I wonder if this is two classes one extends the other with a `protected String getHint()` method, that in the base, it's (HintRequestContext) and the request header specific one you find the header? You'd have to explicitly say you wanted one or the other. Thoughts?

Reply:
IDK, I could maybe see a scenario where you set a hint via properties as the default, and then in some cases use the header to override that default.  Having two separate classes and having to chose one method or the other would make that use case impossible.  

At the very least I think the logic to find the hint to use should be moved to its own methods in this class to clean things up a bit and make the code easier to read.

Reply:
Yes, I agree with @ryanjbaxter  - the "default override" scenario was why I put this all together. Will extract to separate methods to make the code cleaner, though.